vqe.py

```diff
@@ -0,0 +1,84 @@
+import networkx as nx
+import numpy as np
+
+from graphix import Circuit
+from graphix.simulator import PatternSimulator
+
+# Define the Hamiltonian for the VQE problem (Example: H = Z0Z1 + X0 + X1)
+def create_hamiltonian():
+    Z = np.array([[1, 0], [0, -1]])
+    X = np.array([[0, 1], [1, 0]])
+    H = np.kron(Z, Z) + np.kron(X, np.eye(2)) + np.kron(np.eye(2), X)
+    return H
+
+# Function to build the VQE circuit
+def build_vqe_circuit(n_qubits, params):
+    circuit = Circuit(n_qubits)
+    for i in range(n_qubits):
+        circuit.rx(i, params[i])
+        circuit.ry(i, params[i + n_qubits])
+        circuit.rz(i, params[i + 2 * n_qubits])
+    for i in range(n_qubits - 1):
+        circuit.cnot(i, i + 1)
+    return circuit
+
+# Define the class for VQE with MBQC
+class MBQCVQE:
+    def __init__(self, n_qubits, hamiltonian):
+        self.n_qubits = n_qubits
+        self.hamiltonian = hamiltonian
+
+    def build_mbqc_pattern(self, params):
+        circuit = build_vqe_circuit(self.n_qubits, params)
+        pattern = circuit.transpile().pattern
+        pattern.standardize()
+        pattern.shift_signals()
+        return pattern
+
+    def simulate_mbqc(self, params, backend="statevector"):
+        pattern = self.build_mbqc_pattern(params)
+        pattern.perform_pauli_measurements()  # Perform Pauli measurements
+        simulator = PatternSimulator(pattern, backend=backend)
+        if backend == "tensornetwork":
+            tn = simulator.run()  # Simulate the MBQC circuit using tensor network
+            return tn
+        else:
+            out_state = simulator.run()  # Simulate the MBQC circuit using other backends
+            return out_state
+
+    def compute_energy(self, params):
+        # Simulate the statevector using MBQC
+        out_state = self.simulate_mbqc(params)
+        statevec = out_state.psi.flatten()
+        # Compute the expectation value 
+        energy = np.real(np.dot(statevec.conjugate().T, np.dot(self.hamiltonian, statevec)))
+        return energy
+
+# Set parameters for VQE
+n_qubits = 2
+hamiltonian = create_hamiltonian()
+params = np.random.rand(n_qubits * 3)  # Random initial parameters
+
+# Instantiate the MBQCVQE class
+mbqc_vqe = MBQCVQE(n_qubits, hamiltonian)
+
+# Simulate the VQE circuit using MBQC with tensor network backend
```

Comment:
can you perform the actual optimization to obatin the optimum parameters, and check they make sense?

Reply:
I meant, can you use e.g. `scipy.optimize.minimize` to optimize the parameters and see if it works? this requires many runs of MBQCVQ. compute_energy. good example would be [this](https://graphix.readthedocs.io/en/latest/gallery/qnn.html), though this is different application.

Reply:
nice, very close! are the new code being pushed to [AbdullahKazi500:AbdullahKazi500-patch-2](https://github.com/AbdullahKazi500/graphix/tree/AbdullahKazi500-patch-2) branch or somewhere else? could you push to [AbdullahKazi500:AbdullahKazi500-patch-2](https://github.com/AbdullahKazi500/graphix/tree/AbdullahKazi500-patch-2) so that they will be reflected to thie PR?

Reply:
lmk if you run into further problems. I think it might just be simple (though not great practice..) to just paste new code in vqe.py of patch-2 branch and push, since there is just one file being changed.